packages/tql/src/dialects/postgres.ts

```diff
@@ -0,0 +1,104 @@
+import { IdenticalColumnValidator } from '../values-object-validator';
+import type { DialectImpl } from '../types';
+import { BaseDialect } from '../dialect';
+import {
+  type TqlIdentifiers,
+  type TqlList,
+  type TqlParameter,
+  type TqlTemplateString,
+  type TqlSet,
+  type TqlValues,
+} from '../nodes.js';
+
+export class PostgresDialect extends BaseDialect implements DialectImpl {
+  string(str: TqlTemplateString): void {
+    this.appendToQuery(str.value);
+  }
+
+  parameter(param: TqlParameter): void {
+    const paramNumber = this.appendToParams(param.value);
+    this.appendToQuery(`$${paramNumber}`);
+  }
+
+  identifiers(ids: TqlIdentifiers): void {
+    if (Array.isArray(ids.values)) {
+      this.appendToQuery(
+        ids.values.map((v) => PostgresDialect.escapeIdentifier(v)).join(', '),
+      );
+    } else {
+      this.appendToQuery(PostgresDialect.escapeIdentifier(ids.values));
+    }
+  }
+
+  list(vals: TqlList): void {
+    this.appendToQuery('(');
+    const queryItems: string[] = [];
+    for (const param of vals.values) {
+      const paramNumber = this.appendToParams(param);
+      queryItems.push(`$${paramNumber}`);
+    }
+    this.appendToQuery(queryItems.join(', '));
+    this.appendToQuery(')');
+  }
+
+  values(entries: TqlValues): void {
+    if (Array.isArray(entries.values)) {
+      // it's multiple entries
+      const validator = new IdenticalColumnValidator();
+      let first = true;
+      let columns: string[] = [];
+      const rows: string[] = [];
+      for (const entry of entries.values) {
+        validator.validate(entry);
+        if (first) {
+          first = false;
+          columns = Object.keys(entry);
+          this.appendToQuery(
+            `(${columns
+              // eslint-disable-next-line @typescript-eslint/unbound-method -- this rule is idiotic, this is a static method
+              .map(PostgresDialect.escapeIdentifier)
+              .join(', ')}) VALUES `,
+          );
+        }
+        const queryItems: string[] = [];
+        for (const column of columns) {
+          const paramNumber = this.appendToParams(entry[column]);
+          queryItems.push(`$${paramNumber}`);
+        }
+        rows.push(`(${queryItems.join(', ')})`);
+      }
+      this.appendToQuery(rows.join(', '));
+    } else {
+      // it's a single entry
+      const entry = entries.values;
+      const columns = Object.keys(entry);
+      this.appendToQuery(
+        // eslint-disable-next-line @typescript-eslint/unbound-method -- this rule is idiotic, this is a static method
+        `(${columns.map(PostgresDialect.escapeIdentifier).join(', ')}) VALUES `,
+      );
+      const queryItems: string[] = [];
+      for (const column of columns) {
+        const paramNumber = this.appendToParams(entry[column]);
+        queryItems.push(`$${paramNumber}`);
+      }
+      this.appendToQuery(`(${queryItems.join(', ')})`);
```

Comment:
Hmm, you know, that might work in practice, but I think it might grow better if we leave it separate -- I'm not very confident the syntaxes are _always_ the same -- and additionally, we'd have to turn the entry/column map into a `TqlList` before passing it to `this.list`, which probably would end up being more code anyway.